benchmarks/src/main/java/io/grpc/benchmarks/qps/AsyncServer.java

```diff
@@ -184,36 +191,173 @@ public ForkJoinWorkerThread newThread(ForkJoinPool pool) {
+        .build();
+
+    private final AtomicBoolean shutdown = new AtomicBoolean();
+
+    public BenchmarkServiceImpl() {
+    }
+
+    public void shutdown() {
+      shutdown.set(true);
+    }
 
     @Override
-    public void unaryCall(SimpleRequest request, StreamObserver<SimpleResponse> responseObserver) {
-      SimpleResponse response = Utils.makeResponse(request);
-      responseObserver.onNext(response);
+    public void unaryCall(Messages.SimpleRequest request,
+        StreamObserver<Messages.SimpleResponse> responseObserver) {
+      responseObserver.onNext(Utils.makeResponse(request));
       responseObserver.onCompleted();
     }
 
     @Override
-    public StreamObserver<SimpleRequest> streamingCall(
-        final StreamObserver<SimpleResponse> responseObserver) {
-      return new StreamObserver<SimpleRequest>() {
+    public StreamObserver<Messages.SimpleRequest> streamingCall(
+        final StreamObserver<Messages.SimpleResponse> observer) {
+      final ServerCallStreamObserver<Messages.SimpleResponse> responseObserver =
+          (ServerCallStreamObserver<Messages.SimpleResponse>) observer;
+      // TODO(spencerfang): flow control to stop reading when !responseObserver.isReady
+      return new StreamObserver<Messages.SimpleRequest>() {
         @Override
-        public void onNext(SimpleRequest request) {
-          SimpleResponse response = Utils.makeResponse(request);
-          responseObserver.onNext(response);
+        public void onNext(Messages.SimpleRequest value) {
+          if (shutdown.get()) {
+            responseObserver.onCompleted();
+            return;
+          }
+          responseObserver.onNext(Utils.makeResponse(value));
         }
 
         @Override
         public void onError(Throwable t) {
-          System.out.println("Encountered an error in streamingCall");
-          t.printStackTrace();
+          // other side closed with non OK
+          responseObserver.onError(t);
         }
 
         @Override
         public void onCompleted() {
+          // other side closed with OK
           responseObserver.onCompleted();
         }
       };
     }
 
+    @Override
+    public StreamObserver<Messages.SimpleRequest> streamingFromClient(
+        final StreamObserver<Messages.SimpleResponse> responseObserver) {
+      return new StreamObserver<Messages.SimpleRequest>() {
+        Messages.SimpleRequest lastSeen = null;
+
+        @Override
+        public void onNext(Messages.SimpleRequest value) {
+          if (shutdown.get()) {
+            responseObserver.onCompleted();
+            return;
+          }
+          lastSeen = value;
+        }
+
+        @Override
+        public void onError(Throwable t) {
+          // other side closed with non OK
+          responseObserver.onError(t);
+        }
+
+        @Override
+        public void onCompleted() {
+          if (lastSeen != null) {
+            responseObserver.onNext(Utils.makeResponse(lastSeen));
```

Comment:
This would be the case where a shutdown races with an `onCompleted`, and I think either is fine here. The client should be streaming requests to us and `onNext` would have detected the shutdown. I'll change that case to an `onError` to be more clear.